### PR TITLE
Qwen35 sp fix  by ulysses

### PR DESCRIPTION
**Pull request by meichangsu1**
# PR type
- [ x] Bug Fix
- [ ] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

实现思路
本次实现采用的是“通用 SP 策略 + Qwen3.5 模型特化实现”的分层方案。
1. 通用 SP 部分放在 strategy 层
通用的 Ulysses/SP 能力放在 sequence_parallel.py 中，负责：
根据 ulysses_size 在序列维度切分输入
在 attention 内部执行 seq <-> head 的 all-to-all 通信
维护 sequence parallel runtime context等运行时信息

2. Qwen3.5 特有部分放在模型层
Qwen3.5 中使用的是线性注意力，GatedDeltaNet 还有卷积层，仍然沿用ulysses 的思想
参考思路： [`Linear Attention Ulysses (GatedDeltaNet)`](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs/key_features/ulysses.md#-linear-attention-ulysses-gateddeltanet)

只是在 modeling_qwen3_5.py 中，目前重写了这些类
TwinkleQwen3_5TextModel
TwinkleQwen3_5DecoderLayer
TwinkleQwen3_5GatedDeltaNet
TwinkleQwen3_5ForCausalLM

在用户使用时指定model_cls=TwinkleQwen3_5ForCausalLM：
  model = TransformersModel(
        model_id=MODEL_ID,
        model_cls=TwinkleQwen3_5ForCausalLM,
        device_mesh=device_mesh,
        strategy='native_fsdp',
    )

综上
因此 softmax attention 继续复用通用 SP
linear attention 由模型内部显式处理 ，让 Qwen3.5 的语言模型骨干本身具备 SP 感知能力。

## Experiment results
开启sp(ulysses=2)
[2026-03-23 15:31:43][INFO:twinkle] Current is optimizer step 60 of 63 (micro step 121 of 125), metric: {'loss': '1.1285', 'grad_norm': '0.139006', 'accuracy': '0.68', 'correct_tokens': 21312, 'total_tokens': 31280, 'learning rate(param group 1)': '6.586739e-07', 'learning rate(param group 2)': '6.586739e-07', 'iters': 60, 'total time elapse': '187 seconds', 'speed': '0.36 iters/s', 'rank': 0, 'local_rank': 0, 'device': 'cuda:0', 'mem_allocated': '2200.0 MiB', 'mem_reserved': '55992.0 MiB', 'mem_peak_allocated': '27645.0 MiB', 'mem_peak_reserved': '55992.0 MiB'}

关闭sp
[2026-03-23 13:47:27][INFO:twinkle] Current is step 60 of 63, metric: {'loss': '1.1290', 'grad_norm': '0.123370', 'accuracy': '0.68', 'correct_tokens': 21283, 'total_tokens': 31280, 'learning rate(param group 1)': '6.586739e-07', 'learning rate(param group 2)': '6.586739e-07', 'iters': 60, 'total time elapse': '129 seconds', 'speed': '0.53 iters/s', 'rank': 0, 'local_rank': 0, 'device': 'cuda:0', 'mem_allocated': '2183.5 MiB', 'mem_reserved': '62422.0 MiB', 'mem_peak_allocated': '36788.4 MiB', 'mem_peak_reserved': '77750.0 MiB'}

### CODE REVIEW

File: src/twinkle/dataloader/dataloader.py [hunk 1]
```diff
@@ -45,14 +45,31 @@ def __init__(self,
...
+            assert batch_size >= required_world_size and batch_size % required_world_size == 0
+        self.batch_size = self._resolve_runtime_batch_size(batch_size, device_mesh)
         self.dataloader_params = kwargs
-        self.dataloader_params['batch_size'] = batch_size
+        self.dataloader_params['batch_size'] = self.batch_size
         self.device_mesh = device_mesh
         self.processor: Optional[InputProcessor] = None
         self._set_work_init_fn()
 
+    @staticmethod
```
Comment (tastelikefeet) [inline src/twinkle/dataloader/dataloader.py:57]: data_world_size 讲道理应该包含ulysses的判断才对

Comment (tastelikefeet) [reply src/twinkle/dataloader/dataloader.py:57]: world_size的判断放在具体组件里不太合适，收敛到device_mesh中比较好

File: src/twinkle/dataloader/dataloader.py [hunk 2]
```diff
@@ -45,14 +45,31 @@ def __init__(self,
         self.max_retries = kwargs.pop('max_retries', 20)
         self.min_batch_size = min_batch_size
         if device_mesh is not None:
-            assert batch_size >= device_mesh.data_world_size and batch_size % device_mesh.data_world_size == 0
-        self.batch_size = batch_size
+            required_world_size = self._required_data_world_size(device_mesh)
+            assert batch_size >= required_world_size and batch_size % required_world_size == 0
+        self.batch_size = self._resolve_runtime_batch_size(batch_size, device_mesh)
```
Comment (tastelikefeet) [inline src/twinkle/dataloader/dataloader.py:50]: 这里的逻辑是什么原因呢

File: src/twinkle/model/transformers/transformers.py [hunk 3]
```diff
@@ -151,6 +151,13 @@ def calculate_metrics(self, is_training):
 DEFAULT_WEIGHT_DECAY = 0.01
 
 
+def _default_gradient_accumulation_steps_for_device_mesh(device_mesh: Optional[DeviceMesh]) -> int:
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py:154]: 这个是否可以让用户自行决定
